normandy/base/middleware.py

```diff
@@ -128,3 +136,109 @@ def process_request(self, request):
                 }
             )
         return response
+
+
+def default_username(email):
+    """Generate username for the Django user.
+
+    :arg str/unicode email: the email address to use to generate a username
+
+    :returns: str/unicode
+
+    """
+    # bluntly stolen from django-browserid
+    # store the username as a base64 encoded sha224 of the email address
+    # this protects against data leakage because usernames are often
+    # treated as public identifiers (so we can't use the email address).
+    return smart_text(base64.urlsafe_b64encode(
+        hashlib.sha1(force_bytes(email)).digest()
+    ).rstrip(b'='))
+
+
+class OIDCAccessTokenAuthenticationMiddleware:
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+        # XXX Here we *could* do something like:
+        # if not settings.OIDC_ACCESS_TOKEN_ENABLED:
+        #    from django.core.exceptions import MiddlewareNotUsed
+        #    raise MiddlewareNotUsed
+        self.oidc_user_endpoint = settings.OIDC_USER_ENDPOINT
+        self.UserModel = auth.get_user_model()
+
+    def __call__(self, request):
+        response = self.process_request(request)
+        if not response:
+            response = self.get_response(request)
+        return response
+
+    def process_request(self, request):
+        try:
+            access_token = re.findall(
+                'Bearer\s+(.*)',  # Is this right?
+                request.META.get('HTTP_AUTHORIZATION') or '',
+                re.I
+            )[0]
+        except IndexError:
+            # That's cool. If it's not there, it's not there.
+            # This middleware will not be able to authenticate the user
+            # for the duration of this request.
+            return
+        cache_key = 'oidc-access-token-check:{}'.format(
+            hashlib.md5(force_bytes(access_token)).hexdigest()
+        )
+        checked = cache.get(cache_key)
+        if checked:
+            return checked
+        # print("access_token", repr(access_token))
+        # XXX This should use a retrying requests sesssion or backoff lib.
+        response = requests.get(self.oidc_user_endpoint, headers={
+            'Authorization': 'Bearer {0}'.format(access_token)
+        })
+        if response.status_code == 200:
+            # Awesome! Turn this into a Django User
+            user_profile = response.json()
+            if 'email' not in user_profile:
+                raise ImproperlyConfigured(
+                    "The HTTP Authorization header had an access token that "
+                    "was created without the 'email' scope."
+                )
+            email = user_profile['email']
+            if not email:
+                # XXX Not sure how this could happen.
```

Comment:
This is totally possible. If someone does this:
```
new auth0.WebAuth({
    domain: OIDC_DOMAIN,
    clientID: OIDC_CLIENT_ID,
    redirectUri: OIDC_CALLBACK_URL,
    responseType: 'token id_token',
    scope: 'openid profile',  // <--- NOTE! No 'email' scope
  });
```

We should definitely raise a `PermissionDenied` but I don't know if whatever message we put in there will be visible to the client if they debug the response body. Either way, this would be a good opportunity to explain the OIDC scope thing at least.